ps_qualityassurance.php

```diff
@@ -81,26 +102,113 @@ public function __call($methodName, array $arguments)
+
+            echo $output;
+        }
+    }
+
+    public function getContent()
+    {
+        $moduleAdminLink = Context::getContext()->link->getAdminLink('AdminQualityAssurance', true);
+        Tools::redirectAdmin($moduleAdminLink);
+    }
+
+    /**
+     * Look whether there is a payload registered for this hook
+     *
+     * @param string $hookName
+     * @return array|null
+     */
+    protected function getRegisteredHookPayload($hookName)
+    {
         $query = new DbQuery();
         $query->select('content')
-              ->select('enabled');
+            ->select('enabled');
         $query->from('quality_assurance_hooks');
         $query->where('name = "' . pSQL($hookName) . '"');
 
         $row = Db::getInstance()->getRow($query);
-        if (!empty($row['enabled'])) {
-            $params = !empty($arguments[0]) ? $arguments[0] : [];
-            try {
-                return eval($row['content']);
-            } catch (Throwable $e) {
-                echo $e->getMessage();
+
+        return $row;
+    }
+
+    /**
+     * @param string $hookName
+     * @param array $arguments
+     * @param string|null $output
+     * @param bool|null $error
+     *
+     * @return bool
+     */
+    protected function recordHookCall(string $hookName, array $arguments, ?string $output = null, ?bool $error = false)
+    {
+        $requestIdentifier = $this->getRequestIdentifier();
+
+        $parameters = sprintf("('%s', '%s', '%s', '%s', now(), %d)",
+            pSQL($requestIdentifier),
+            pSQL($hookName),
+            $this->makeArgumentsReadable($arguments),
+            sprintf('%s: %s',
+                (!$error ? 'Output' : 'Error'),
+                pSQL($output)
+            ),
+            ($error ? 1 : 0)
+        );
+
+        $sql = 'INSERT INTO ' . _DB_PREFIX_ . 'quality_assurance_hook_logs (request_identifier, hook_name, hook_parameters, output, called_at, error) VALUES '
+            . $parameters;
+
+        return Db::getInstance()->execute($sql);
+    }
+
+    /**
+     * @param array $arguments
+     *
+     * @return string
+     */
+    protected function makeArgumentsReadable(array $arguments)
+    {
+        $result = [];
+        foreach ($arguments as $key => $value) {
+            if (is_array($value)) {
+                $result[$key] = 'Array';
+            } elseif (is_object($value)) {
+                $result[$key] = sprintf('Object %s', get_class($value));
+            } else {
+                $result[$key] = (string)$value;
             }
         }
+
+        return json_encode($result);
     }
 
-    public function getContent()
+    /**
+     * @return string
+     */
+    protected function getRequestIdentifier()
     {
-        $moduleAdminLink = Context::getContext()->link->getAdminLink('AdminQualityAssurance', true);
-        Tools::redirectAdmin($moduleAdminLink);
+        if ($this->requestIdentifier === null) {
+            $this->requestIdentifier = random_bytes(20);
```

Comment:
I don't see why you use a requestIdentifier when you have an id in the table :thinking:

Reply:
I want to group together hook calls performed in the same request->response process.
So if a HTTP request hits  `/homepage` endpoint and it triggers 4 hook calls I want to be able to connect these 4 hook calls together. This helps the QA analyst to see "all hook calls triggered by my latest interaction with PrestaShop".

Reply:
Oh, I understand now. But prever uniqid() instead of random_bytes :)

Reply:
Fixed in https://github.com/PrestaShop/ps_qualityassurance/pull/8/commits/da03ba8a722b9e6ec2da896b72f65667ddc7c0a4